File: training-data/cpp/gnome_sort.cpp
#include <functional>
#include <iostream>
#include <string>
#include <vector>

template <typename T, typename Compare = std::less<T>>
void gnome_sort(std::vector<T>& a, Compare cmp = Compare{}) {
    std::size_t i = 0;
    while (i < a.size()) {
        if (i == 0 || !cmp(a[i], a[i - 1])) {
            ++i;
        } else {
            std::swap(a[i], a[i - 1]);
            --i;
        }
    }
}

int main() {
    std::vector<int> nums{34, 2, 10, -9, 7};
    gnome_sort(nums, std::greater<>{});
    for (int x : nums) std::cout << x << ' ';
    std::cout << '\n';
    std::vector<std::string> words{"delta", "alpha", "charlie", "bravo"};
    gnome_sort(words);
    for (const auto& w : words) std::cout << w << ' ';
    std::cout << '\n';
}

File: training-data/rust/gnome_sort.rs
fn gnome_sort_by<T, F: Fn(&T, &T) -> bool>(a: &mut [T], less: F) {
    let mut i = 0;
    while i < a.len() {
        if i == 0 || !less(&a[i], &a[i - 1]) {
            i += 1;
        } else {
            a.swap(i, i - 1);
            i -= 1;
        }
    }
}

fn main() {
    let mut nums = vec![34, 2, 10, -9, 7];
    gnome_sort_by(&mut nums, |a, b| a < b);
    println!("{:?}", nums);
    let mut words = vec!["Delta", "alpha", "Charlie", "bravo"];
    gnome_sort_by(&mut words, |a, b| a.to_lowercase() < b.to_lowercase());
    println!("{:?}", words);
}
